pkg/kubelet/cm/memorymanager/memory_manager.go

```diff
@@ -0,0 +1,408 @@
+func (m *manager) Allocate(pod *v1.Pod, container *v1.Container) error {
+	// Garbage collect any stranded resources before allocation
+	m.removeStaleState()
+
+	m.Lock()
+	defer m.Unlock()
+
+	// Call down into the policy to assign this container memory if required.
+	if err := m.policy.Allocate(m.state, pod, container); err != nil {
+		klog.Errorf("[memorymanager] Allocate error: %v", err)
+		return err
+	}
+	return nil
+}
+
+// RemoveContainer removes the container from the state
+func (m *manager) RemoveContainer(containerID string) error {
+	m.Lock()
+	defer m.Unlock()
+
+	// if error appears it means container entry already does not exist under the container map
+	podUID, containerName, err := m.containerMap.GetContainerRef(containerID)
+	if err != nil {
+		klog.Warningf("[memorymanager] Failed to get container %s from container map error: %v", containerID, err)
+		return nil
+	}
+
+	err = m.policyRemoveContainerByRef(podUID, containerName)
+	if err != nil {
+		klog.Errorf("[memorymanager] RemoveContainer error: %v", err)
+		return err
+	}
+
+	return nil
+}
+
+// State returns the state of the manager
+func (m *manager) State() state.Reader {
+	return m.state
+}
+
+// GetPodTopologyHints returns the topology hints for the topology manager
+func (m *manager) GetPodTopologyHints(pod *v1.Pod) map[string][]topologymanager.TopologyHint {
+	// Garbage collect any stranded resources before providing TopologyHints
+	m.removeStaleState()
+	// Delegate to active policy
+	return m.policy.GetPodTopologyHints(m.state, pod)
+}
+
+// GetTopologyHints returns the topology hints for the topology manager
+func (m *manager) GetTopologyHints(pod *v1.Pod, container *v1.Container) map[string][]topologymanager.TopologyHint {
+	// Garbage collect any stranded resources before providing TopologyHints
+	m.removeStaleState()
+	// Delegate to active policy
+	return m.policy.GetTopologyHints(m.state, pod, container)
+}
+
+func (m *manager) removeStaleState() {
+	// Only once all sources are ready do we attempt to remove any stale state.
+	// This ensures that the call to `m.activePods()` below will succeed with
+	// the actual active pods list.
+	if !m.sourcesReady.AllReady() {
+		return
+	}
+
+	// We grab the lock to ensure that no new containers will grab memory block while
+	// executing the code below. Without this lock, its possible that we end up
+	// removing state that is newly added by an asynchronous call to
+	// AddContainer() during the execution of this code.
+	m.Lock()
+	defer m.Unlock()
+
+	// Get the list of active pods.
+	activePods := m.activePods()
+
+	// Build a list of (podUID, containerName) pairs for all containers in all active Pods.
+	activeContainers := make(map[string]map[string]struct{})
+	for _, pod := range activePods {
+		activeContainers[string(pod.UID)] = make(map[string]struct{})
+		for _, container := range append(pod.Spec.InitContainers, pod.Spec.Containers...) {
```

Comment:
can we leave a placeholder in this interface for ephemeral containers?  just something that notes its needed, or maybe error if ephemeral containers are enabled?

Reply:
Can you please clarify what should be done here regarding the ephemeral containers? From the documentation, I can see that
```
Ephemeral containers differ from other containers in that they lack guarantees for resources or execution, and they will never be automatically restarted, so they are not appropriate for building applications.
```
so in general we do not allocate any memory to it during the allocation phase and we also do not need to remove it from the state during call to the `removeStaleState`

Reply:
@derekwaynecarr 
This function is responsible for garbage collecting any stale state in the `MemoryManager` for containers that have exited but not been cleaned up yet.

There should never be a case where ephemeral containers have an entry in this state. The have been allocated `MemoryManager` only allocates memory to containers from pods in the `GuaranteedQoS` class and only if they have `requested==limits`. Ephemeral containers should never fall into this case.